### PR TITLE
Make -frewrite-includes put an endif at the end of the included text

### PR DESCRIPTION
**Pull request by pogo59**
Also add the filename to the comments it emits, to help identify where included text ends.

### CODE REVIEW

File: clang/lib/Frontend/Rewrite/InclusionRewriter.cpp
Comment (erichkeane) [inline clang/lib/Frontend/Rewrite/InclusionRewriter.cpp:349]: I DO think there is at least a good attempt that could be made to get a unique set of names for each (perhaps a stringify of the path, ala how `#line` does it?), but I think this is a good stand-alone feature here.

One thing I MIGHT suggest anyway though is replacing the '0' with `defined(__CLANG_REWRITTEN_INCLUDES)`, so that turning on the `#include` version would be trivial.

I might also suggest for certain lines, `defined(__CLANG_REWRITTEN_SYSTEM_INCLUDES)` such that only the system headers are used from the test machine.

So that is, a system header would be:

```#if defined(__CLANG_REWRITTEN_INCLUDES) || defined(__CLANG_REWRITTEN_SYSTEM_INCLUDES) /* vector expanded by -frewrite-includes */```

**Review by erichkeane — COMMENTED** [review]
From what I can tell, this looks correct, though I don't know much about this rewriter, I've only spent a little time reviewing it.  That said, i think this is a very nice improvement.

One thing I might request (in addition to what I did in teh comments :D) is a similar change in `-frewrite-modules`.

File: clang/test/Modules/preprocess-module.cpp
Comment (pogo59) [reply clang/test/Modules/preprocess-module.cpp:101]: Yes it does, but the point here is simply to identify that this is the correct `#else`. The Preprocessor tests look for the full comment.

File: clang/lib/Frontend/Rewrite/InclusionRewriter.cpp
Comment (pogo59) [reply clang/lib/Frontend/Rewrite/InclusionRewriter.cpp]: TIL about `#pragma clang system_header` which turns a header into a system header, regardless of how it was found. But, by the time the rewriter sees the pragma, it has already emitted the `#include` and its wrapper `#if`. That makes the `__CLANG_REWRITTEN_SYSTEM_INCLUDES` not as reliable, and I didn't do that part (yet).
But, I did do the `defined(__CLANG_REWRITTEN_INCLUDES)` part, so you can get all-or-nothing.

Comment (erichkeane) [reply clang/lib/Frontend/Rewrite/InclusionRewriter.cpp]: Hmm... ok.  The 'all or nothing' is fine.  As far as the system-includes, I wouldn't mind it if we were imperfect, as this is exclusively a debugging tool, thus any amount of ABI/API/etc stability doesn't really exist.  So I'm ok with a bit of a 'best effort' here (same on the enabling individual headers).  

we can always revise this as we determine needs.

BTW, I'm ok with this patch, but I am not the code owner here, so I'd like to give others a better chance to review this, particularly @MaskRay .

Comment (pogo59) [reply clang/lib/Frontend/Rewrite/InclusionRewriter.cpp]: Aha, `-MM` has the same problem, so I'm less concerned about that now. I'll add the toggle for "system" headers later today.

**Comment by pogo59** [comment]
Toggles done. Also got rid of some of the extra whitespace changes in the output.

**Review by erichkeane — COMMENTED** [review]
This looks great/useful to me.  I have no problems with it, but would like to give this a bit of time for others to review.  If this doesn't get other comments by friday next week, I'll approve.

**Comment by danakj** [comment]
I was just lamenting the lack of this feature this week. Thank you, I can’t wait to use it.

**Comment by pogo59** [comment]
Ping

**Review by erichkeane — COMMENTED** [review]
The clang-format tool still has a complaint, so this isn't passing CI.  Unfortunately it doesn't seem to print it in the CI, so you'll have to run it locally.  Delta that, LGTM.

**Review by erichkeane — APPROVED** [review]
Approved, assuming this passes CI.

**Pull request merged**
